sql/src/main/java/org/apache/druid/sql/calcite/external/TableAppendMacro.java

```diff
@@ -0,0 +1,309 @@
+    AppendDatasourceMetadata metadata = buildUnionDataSource(tables);
+    return new DatasourceTable(
+        metadata.values,
+        metadata,
+        EffectiveMetadata.of(metadata.values)
+    );
+  }
+
+  private List<RelOptTable> getTables(SqlOperatorBinding callBinding, SqlValidatorCatalogReader catalogReader)
+  {
+    List<RelOptTable> tables = new ArrayList<>();
+    for (int i = 0; i < callBinding.getOperandCount(); i++) {
+      if (!callBinding.isOperandLiteral(i, false)) {
+        throw new IllegalArgumentException(
+            "All arguments of call to macro "
+                + "APPEND should be literal. Actual argument #"
+                + i + " is not literal"
+        );
+      }
+      @Nullable
+      String tableName = callBinding.getOperandLiteralValue(i, String.class);
+      ImmutableList<String> tableNameList = ImmutableList.<String>builder().add(tableName).build();
+      SqlValidatorTable table = catalogReader.getTable(tableNameList);
+      if (table == null) {
+        throw DruidSqlValidator.buildCalciteContextException(
+            StringUtils.format("Table [%s] not found", tableName),
+            ((SqlCallBinding) callBinding).operand(i)
+        );
+      }
+      tables.add(table.unwrapOrThrow(RelOptTable.class));
+    }
+    return tables;
+  }
+
+  static class AppendDatasourceMetadata implements DatasourceMetadata
+  {
+    private final RowSignature values;
+    private final DataSource dataSource;
+
+    public AppendDatasourceMetadata(RowSignature values, List<DataSource> dataSources)
+    {
+      this.values = values;
+      this.dataSource = new UnionDataSource(dataSources);
+    }
+
+    @Override
+    public boolean isJoinable()
+    {
+      return false;
+    }
+
+    @Override
+    public boolean isBroadcast()
+    {
+      return false;
+    }
+
+    @Override
+    public DataSource dataSource()
+    {
+      return dataSource;
+    }
+  }
+
+  private AppendDatasourceMetadata buildUnionDataSource(List<RelOptTable> tables)
+  {
+    List<DataSource> dataSources = new ArrayList<>();
+    Map<String, ColumnType> fields = new LinkedHashMap<>();
+    Builder rowSignatureBuilder = RowSignature.builder();
+    for (RelOptTable relOptTable : tables) {
+      DatasourceTable table = relOptTable.unwrapOrThrow(DatasourceTable.class);
+      RowSignature rowSignature = table.getRowSignature();
+      for (String columnName : rowSignature.getColumnNames()) {
+        ColumnType currentType = rowSignature.getColumnType(columnName).get();
+        ColumnType existingType = fields.get(columnName);
+
+        if (existingType == null || existingType.equals(currentType)) {
+          fields.put(columnName, currentType);
+        } else {
+          try {
+            ColumnType commonType = ColumnType.leastRestrictiveType(currentType, existingType);
+            fields.put(columnName, commonType);
+          }
+          catch (Exception e) {
+            throw InvalidInput.exception(
```

Comment:
`InvalidSqlInput`?

Reply:
If we don't except this Exception to ever happen, then maybe make it a `DruidException.defensive()` instead and perhaps add something to the message about why it should never occur.